File: June-LeetCoding-Challenge-2020/Week1/528-random-pick-with-weight.py
class Solution:
    def __init__(self, w: List[int]):
        self.part_weight = []
        total = 0
        for weight in w:
            total += weight
            self.part_weight.append(total)
        self.total_weight = total

    def pickIndex(self) -> int:
        ran = random.randint(1,self.total_weight)
        for index, weight in enumerate(self.part_weight):
            if ran <= weight:
                return index

# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()
    # TC: O(n)
    # SC: O(n)
    # Runtime: 5760 ms, faster than 8.55%, might need to work
    # Memory Usage: 18.3 MB, less than 50.86%

File: June-LeetCoding-Challenge-2020/Week1/528-random-pick-with-weight.cpp
// Time Complexity: O(n)
// Space Complexiy: O(n)
// Runtime: 796 ms, faster than 5.26%
// Memory Usage: 40.6 MB, less than 32.99% 
class Solution {
public:
    vector<int> temp;
    int total = 0;
    
    Solution(vector<int>& w) {
        
        temp = w;
        for(int i = 0; i < temp.size(); i++){
            total+=temp[i];
        }
    }
    
    int pickIndex() {
        
        int num = rand()%total + 1;
        int res;
        
        for(int i = 0; i < temp.size(); i++){
            num -= temp[i];
            if(num <= 0)
                return i;
        }
        
        return temp.size()-1;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */